Extend line endings check

Changes in pull requests are checked for proper line endings: LF for most text files or CR/LF for some Windows-specific text files.

I propose that we extend those checks to ensure that each added or modified line has no trailing whitespace or unnecessary blank lines at EOF.

Since the inception of this project I have yet to see a legitimate need for trailing whitespace in any file. On the other hand, trailing whitespace can lead to some difficult to recognize problems.

Consider this sequence in [jvmtiMethod.c](https://github.com/eclipse-openj9/openj9/blob/master/runtime/jvmti/jvmtiMethod.c#L750-L756):
```
#define FLIP_32BIT(index) \
	do { \
		bytecodes[(index) + 0] = j9Bytecodes[(index) + 3]; \
		bytecodes[(index) + 1] = j9Bytecodes[(index) + 2]; \
		bytecodes[(index) + 2] = j9Bytecodes[(index) + 1]; \
		bytecodes[(index) + 3] = j9Bytecodes[(index) + 0]; \
	} while(0)
```
If the first line had trailing whitespace (after the `\`), `gcc` would complain:
```
jvmtiMethod.c:750:27: error: backslash and newline separated by space [-Werror]
 #define FLIP_32BIT(index) \
```
On the other hand, if a space was present after the second of these lines in [OpenJ9.gmk](https://github.com/ibmruntimes/openj9-openjdk-jdk11/blob/openj9/closed/OpenJ9.gmk#L433-L434):
```
	+BOOT_JDK=$(BOOT_JDK) $(EXPORT_COMPILER_ENV_VARS) OPENJDK_VERSION_NUMBER_FOUR_POSITIONS=$(VERSION_NUMBER_FOUR_POSITIONS) \
		$(MAKE) $(MAKE_ARGS) -C $(OUTPUTDIR)/vm -f $(OPENJ9_TOPDIR)/runtime/buildtools.mk \
```
an UMA build would fail:
```
cc      -o .o
cc: fatal error: no input files
compilation terminated.
<builtin>: recipe for target '.o' failed
```
Not many people will successfully trace that back to the offending space.

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -68,6 +68,9 @@ Following these guidelines will help us merge your pull requests smoothly:
 5. Please carefully read and adhere to the legal considerations and
    copyright/license requirements outlined below.
 
+6. Ensure your changes are compatible with the checks that will be applied to your pull request.
+   - The second year in each copyright notice should be the current year.
+   - Text files should use the proper line-endings and there should be no unwanted whitespace. You can enable the sample `pre-commit` hook created by `git init` to ensure you adhere to expectations.
 
 ## Building and testing
 

--- a/buildenv/jenkins/jobs/infrastructure/lineEndingsCheck.groovy
+++ b/buildenv/jenkins/jobs/infrastructure/lineEndingsCheck.groovy
@@ -97,7 +97,17 @@ timeout(time: 6, unit: 'HOURS') {
                         echo "${HASHES}"
                         sh 'exit 1'
                     } else {
-                        echo "All modified files appear to have correct line endings"
+                        println "Checking for added trailing whitespace..."
+                        sh 'git config core.whitespace blank-at-eof,blank-at-eol,cr-at-eol,space-before-tab'
+                        WHITESPACE_ERRORS = sh (
+                            script: "git diff --check origin/${ghprbTargetBranch} HEAD",
+                            returnStdout: true
+                        )
+                        if (WHITESPACE_ERRORS.trim() == "") {
+                            echo "All modified files appear to have correct line endings"
+                        } else {
+                            echo "${WHITESPACE_ERRORS}"
+                        }
                     }
                 }
             }

